src/compiler/checker.ts

```diff
@@ -43948,6 +43948,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         function helper(condExpr: Expression, body: Expression | Statement | undefined) {
             const location = isLogicalOrCoalescingBinaryExpression(condExpr) ? skipParentheses(condExpr.right) : condExpr;
+
+            switch (condExpr.kind) {
+                case SyntaxKind.RegularExpressionLiteral:
+                case SyntaxKind.FunctionExpression:

```

Comment:
Aware this is still a draft, but function expressions will almost certainly be a mistyped IIFE, so it might be helpful to separate out a specific error message for these &ndash; maybe some sort of "Did you mean to call..." message, à la 2774?

Reply:
`ArrowFunction` should probably also be in this list.